Zend/zend_exceptions.stub.php

```diff
@@ -119,3 +119,26 @@ class ArithmeticError extends Error
 class DivisionByZeroError extends ArithmeticError
 {
 }
+
+final class StackFrame
+{
+    public static function getTrace(int $options = DEBUG_BACKTRACE_PROVIDE_OBJECT, int $limit = 0): array {}
```

Comment:
can the same method prototype be added for `Throwable` interface as well?

Reply:
@mvorisek No it can't. `Throwable` don't decide on how to collect stack trace - it's being populated on every exception object creation thus limiting it on `getTrace()` doesn't change frames collection process. 
It is also not possible to provide objects on exception trace since they would hold a reference which in case of exception is being thrown inside ctor would block calling dtor.

Reply:
> @mvorisek No it can't. `Throwable` don't decide on how to collect stack trace - it's being populated on every exception object creation thus limiting it on `getTrace()` doesn't change frames collection process.

yes, I meant require the new prototype for all classes implementing `Throwable`, populate object as default and return/convert them to an array (like now) based on the `$options`. I think this is possible and it will largely improve the debugging possibilities with object/closure present.

> It is also not possible to provide objects on exception trace since they would hold a reference which in case of exception is being thrown inside ctor would block calling dtor.

for exceptions in ctor, we might simply not populate the object for the 1st `StackFrame` (which makes sense, as the object is not fully constructed)

Reply:
@mvorisek all `Throwable` objects already have a property with trace array, therefore you can enable access to it by reflection and manipulate to suit your preferences. It's not that `getTrace()` builds something it's a simple getter for an array which is accessible from the private scope and it really does nothing more (see https://github.com/php/php-src/blob/master/Zend/zend_exceptions.c#L429-L438 ).

For ctor throwing exception, it's not that obvious, you can nest them again so reducing it for one first frame doesn't solve the problem (see https://bugs.php.net/bug.php?id=45351).

As I wrote before what I can propose is to expose a WeakRef instead of an object instance. IIRC this would not affect refcounting and would allow reaching the object from exception *if it still exists* without changing current PHP behaviour.

Reply:
> @mvorisek all `Throwable` objects already have a property with trace array, therefore you can enable access to it by reflection and manipulate to suit your preferences. It's not that `getTrace()` builds something it's a simple getter for an array which is accessible from the private scope and it really does nothing more (see https://github.com/php/php-src/blob/master/Zend/zend_exceptions.c#L429-L438 ).

`getTrace()` impl. can change and stay fully BC

> For ctor throwing exception, it's not that obvious, you can nest them again so reducing it for one first frame doesn't solve the problem (see https://bugs.php.net/bug.php?id=45351).
> 
> As I wrote before what I can propose is to expose a `WeakRef` instead of an object instance. IIRC this would not affect refcounting and would allow reaching the object from exception _if it still exists_ without changing current PHP behaviour.

It makes sense - use weak ref internally and return object or null (if null originally or weak ref was released)

Other approach migh be to not populate not fully constructed object at all (even if constructor might complete later without an exception).

I think the 2nd approach is more consistent, consider these usecases:
```
funtion run() {
  $a = new class() {
    public function __construct() {
      throw new Exception();
    }
  };
}

try {
  run();
} catch (Exception $e) {
  $obj = $e->getTrace(DEBUG_BACKTRACE_PROVIDE_OBJECT)[0]->getObject();
  // both approaches will/should return null
}
```

```
funtion run() {
  $a = new class() {
    public function run() {
      throw new Exception();
    }
  };
  $a->run();
}

try {
  run();
} catch (Exception $e) {
  $obj = $e->getTrace(DEBUG_BACKTRACE_PROVIDE_OBJECT)[0]->getObject();
  // WeakRef impl. will/might return null because `$a` is no longer referenced
}
```